Fix ARM Errors and update the swagger files.

ericshape opened the pull request:
Fix some ARM Errors and update the swagger files.
The error list is here: https://github.com/Azure/azure-rest-api-specs/pull/12107/checks?check_run_id=1594037578

<i>MSFT employees can try out our new experience at <b>[OpenAPI Hub](https://aka.ms/openapiportal) </b> - one location for using our validation tools and finding your workflow. 
</i>

### Changelog
Please ensure to add changelog with this PR by answering the following questions.
  1. What's the purpose of the update?    
      - [ ] new service onboarding 
      - [ ] new API version 
      - [ ] update existing version for new feature 
      - [x] update existing version to fix swagger quality issue in s360
      - [ ] Other, please clarify 
  2. When you are targeting to deploy new service/feature to public regions? Please provide date, or month to public if date is not available yet.
  3. When you expect to publish swagger? Please provide date, or month to public if date is not available yet.
  4. If it's an update to existing version,  please select SDKs of specific language and CLIs that require refresh after swagger is published.
      - [ ] SDK of .NET (need service team to ensure code readiness)
      - [ ] SDK of Python
      - [ ] SDK of Java
      - [ ] SDK of Js
      - [ ] SDK of Go
      - [ ] PowerShell
      - [ ] CLI
      - [ ] Terraform
      - [ ] No, no need to refresh for updates in this PR

### Contribution checklist:
- [x] I commit to follow the [Breaking Change Policy](http://aka.ms/bcforapi) of “no breaking changes
- [x] I have reviewed the [documentation](https://aka.ms/ameonboard) for the workflow.
- [x] [Validation tools](https://aka.ms/swaggertools) were run on swagger spec(s) and errors have all been fixed in this PR. [How to fix?](https://aka.ms/ci-fix)

If any further question about AME onboarding or validation tools, please view the [FAQ](https://aka.ms/faqinprreview).

### ARM API Review Checklist
- [ ] Ensure to check this box if one of the following scenarios meet updates in the PR, so that label “WaitForARMFeedback” will be added automatically to involve ARM API Review. Failure to comply may result in delays for manifest application. Note this does not apply to data plane APIs, all “removals” and “adding a new property” no more require ARM API review.
  - Adding new API(s)
  - Adding a new API version
  - Adding a new service

- [x] Please ensure you've reviewed following [guidelines](https://aka.ms/rpguidelines) including [ARM resource provider contract](https://github.com/Azure/azure-resource-manager-rpc) and [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md). Estimated time (4 hours). This is required before you can request review from ARM API Review board.

- [ ] If you are blocked on ARM review and want to get the PR merged with urgency, please get the ARM oncall for reviews (*RP Manifest Approvers* team under <ins>Azure Resource Manager service</ins>) from IcM and reach out to them. 

### Breaking Change Review Checklist 
If there are following updates in the PR, ensure to request an approval from API Review Board as defined in the [Breaking Change Policy](http://aka.ms/bcforapi). 

- [ ] Removing API(s) in stable version
- [ ] Removing properties in stable version
- [ ] Removing API version(s) in stable version
- [ ] Updating API in stable version with Breaking Change Validation errors
- [ ] Updating API(s) in preview over 1 year

**I have provided a detailed explanation of breaking changes in the thread.** 

**Action**: to initiate an evaluation of the breaking change, create a new intake using the [template for breaking changes](https://aka.ms/Breakingchangetemplate). Addition details on the process and office hours are on the [Breaking change Wiki](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37684/Breaking-Changes).

Please follow the link to find more details on [PR review process](https://aka.ms/SwaggerPRReview).

ericshape committed:
Fix ARM Errors and update the swagger files.

ericshape commented:
**Notes about the breaking change:**

1. These breaking changes are expected. We reverted previous breaking changes of import/export spec file which are introduced in late Nov 2020. For more detail, please look at this internal PR for more detail:
https://msdata.visualstudio.com/Database%20Systems/_git/DsMainDev/pullrequest/497586 

Rule | Message
-- | --
❌ 1005 - RemovedPath | The new version is missing a path that was found in the old version. Was path '/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Sql/servers/{serverName}/import' removed or restructured?New: azure-rest-api-specs/blob/d44421cb1c524812db4820f4e3096d068d95daa6/Old: Microsoft.Sql/preview/2020-02-02-preview/Databases.json#L660:5
❌ 1005 - RemovedPath | The new version is missing a path that was found in the old version. Was path '/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Sql/servers/{serverName}/import' removed or restructured?New: azure-rest-api-specs/blob/d44421cb1c524812db4820f4e3096d068d95daa6/Old: Microsoft.Sql/preview/2020-08-01-preview/Databases.json#L711:5
❌ 1006 - RemovedDefinition | The new version is missing a definition that was found in the old version. Was 'ImportNewDatabaseDefinition' removed or renamed?New: Microsoft.Sql/preview/2020-02-02-preview/Databases.json#L761:3Old: Microsoft.Sql/preview/2020-02-02-preview/Databases.json#L816:3
❌ 1006 - RemovedDefinition | The new version is missing a definition that was found in the old version. Was 'ImportNewDatabaseDefinition' removed or renamed?New: Microsoft.Sql/preview/2020-08-01-preview/Databases.json#L770:3Old: Microsoft.Sql/preview/2020-08-01-preview/Databases.json#L825:3
❌ 1038 - AddedPath | The new version is adding a path that was not found in the old version.New: Microsoft.Sql/preview/2020-02-02-preview/Servers.json#L19:5
❌ 1038 - AddedPath | The new version is adding a path that was not found in the old version.New: Microsoft.Sql/preview/2020-08-01-preview/Servers.json#L280:5

2. These two breaking changes are also expected. We fix ❌ R2055 - OneUnderscoreInOperationId	ARM error in SQL DB repo. 

Rule | Message
-- | --
❌ 1008 - ModifiedOperationId | The operation id has been changed from 'ManagedDatabaseQueries_ListByQueryId_Get' to 'ManagedDatabaseQueries_Get'. This will impact generated code.New: Microsoft.Sql/preview/2020-02-02-preview/ManagedDatabaseQueries.json#L20:7Old: Microsoft.Sql/preview/2020-02-02-preview/ManagedDatabaseQueries.json#L20:7
❌ 1008 - ModifiedOperationId | The operation id has been changed from 'ManagedDatabaseQueries_ListByQueryId_Get' to 'ManagedDatabaseQueries_Get'. This will impact generated code.New: Microsoft.Sql/preview/2020-08-01-preview/ManagedDatabaseQueries.json#L20:7Old: Microsoft.Sql/preview/2020-08-01-preview/ManagedDatabaseQueries.json#L20:7

ericshape commented:
We also added ServerDevOpsAudit spec and example files to 2020-02-02 folder. They are some missing files which are expected to add last month. This change has been approved by the owner @eswatad

sredman reviewed (approved):
Approving Import/Export service-related changes

ericshape committed:
update

filizt commented on specification/sql/resource-manager/Microsoft.Sql/preview/2020-02-02-preview/ManagedInstances.json:
```diff
@@ -88,7 +88,7 @@
             }
           },
           "default": {
```
>"default": { [](start = 10, length = 12)

The "default" response should have a associated schema.

ericshape replied:
@filizt 
do we have ARM/API/SDK rules to enforce it? All these spec and example files are auto-generated by SQL DB repo. And this issue was not mentioned in any previous PRs.

akning-ms replied:
Hi @ericshape , actually we have ARM Lint rule for this: https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/openapi-authoring-automated-guidelines.md#r4007-defaulterrorresponseschema

But currently to keep your service consistent and breaking change consideration, this Rule is only for new Service, guess why Filizt said ideally. :)

ericshape replied:
Thanks for your comments, Arthur! 
It seems the static validation service is mandatory for SQL DB repo. It will enforce these rules on Day one and reduce the engineer efforts on both ARM/API/SDK and SQL DB sides.

ericshape commented:
❌ RESPONSE_BODY_NOT_IN_EXAMPLE | Response statusCode "200" for operation "DatabaseExtensions_Get" has no response body provided in the example, however the response does have a "schema" defined in the swagger spec.

for this validation error, I will contact the owner in SQL org to fix it soon. 
If it cannot be fixed within this week, I will remove the DatabaseExtension spec and example files from this PR.

akning-ms commented on specification/sql/resource-manager/Microsoft.Sql/preview/2020-02-02-preview/ReplicationLinks.json:
```diff
@@ -124,44 +124,6 @@
             "$ref": "./examples/ReplicationLinkGet.json"
           }
         }
-      },
```
don't suggest to remove the operation if it is for fixing R4011- DeleteOperationResponses.  you can add missing response code to fix it. removing a operation is serious breaking change, not only for SDK.  let us know if there is other reason

akning-ms commented on specification/sql/resource-manager/Microsoft.Sql/preview/2020-02-02-preview/ReplicationLinks.json:
```diff
@@ -124,44 +124,6 @@
         }
-      },
-      "delete": {
-        "tags": [
```
don't suggest to remove the operation if it is for fixing R4011- DeleteOperationResponses. you can add missing response code to fix it. removing a operation is serious breaking change, not only for SDK. let us know if there is other reason

ericshape replied:
Removing the ReplicationLinks_Delete method from V20171001 is the decision from the owner @dlemMSFT and he has approved this PR. Please refer to this internal PR for the detail: https://msdata.visualstudio.com/Database%20Systems/_git/DsMainDev/pullrequest/496954?_a=files

There are two reasons we can safely remove this endpoint from the latest version:
1. there is one ReplicationLinks_Delete method in 2014-04-01 stable version, The SQL API router will auto re-direct the delete request to 2014-04-01 stable. At the same time, @dlemMSFT has confirmed this would not lead any issues. 
2. ReplicationLinks_Delete method from V20171001 was just introduced in Nov 2020. https://github.com/Azure/azure-rest-api-specs/search?q=ReplicationLinks_Delete . Via the search result, we know that the ReplicationLinks_Delete method from V20171001 has not been added to any SDK and doc. v4 SDK still uses the ReplicationLinks_Delete method from 2014-04-01 stable version.

akning-ms replied:
I see, but if we call ReplicationLinks_Delete via version 2020-08-01-preview. it will cause Swagger KPI completeness issue again as before. right?

ericshape replied:
> if we call ReplicationLinks_Delete via version 2020-08-01-preview. it will cause Swagger KPI completeness issue again as before. right?
yes

akning-ms replied:
Hi @ericshape, will you revert this change? I mean it will cause KPI completeness again if you remove this operation

ericshape replied:
yes. we will remove this operation. 'cause we have removed the impl of this API from SQL DB repo.

akning-ms commented:
Noticed pythons sdk have a breaking change caused by remove import_datebase: 
cmdout	[Changelog] **Breaking changes**
cmdout	[Changelog]   - Removed operation DatabasesOperations.import_database
warn	Breaking change found in changelog

Guess due to a python SDK release at end of Dec 2020 .  @msyyc to confirm there is any impact for this removing?

ericshape commented:
> Noticed pythons sdk have a breaking change caused by remove import_datebase:
> cmdout [Changelog] **Breaking changes**
> cmdout [Changelog] - Removed operation DatabasesOperations.import_database
> warn Breaking change found in changelog
> 
> Guess due to a python SDK release at end of Dec 2020 . @msyyc to confirm there is any impact for this removing?

do we release any v4 python SDK at the end of December? I remember we did not release it since there is a no-fly zone in SQL DB from Dec 19 to Jan 10.

akning-ms commented:
> > Noticed pythons sdk have a breaking change caused by remove import_datebase:
> > cmdout [Changelog] **Breaking changes**
> > cmdout [Changelog] - Removed operation DatabasesOperations.import_database
> > warn Breaking change found in changelog
> > Guess due to a python SDK release at end of Dec 2020 . @msyyc to confirm there is any impact for this removing?
> 
> do we release any v4 python SDK at the end of December? I remember we did not release it since there is a no-fly zone in SQL DB from Dec 19 to Jan 10.

Sorry, not end of Dec, accurate time is around 14 dec, there is Python SDK released guess by mistake

jorgecotillo commented on specification/sql/resource-manager/Microsoft.Sql/preview/2020-02-02-preview/ServerDevOpsAudit.json:
```diff
@@ -0,0 +1,277 @@
+            "$ref": "#/parameters/ServerNameParameter"
+          },
+          {
+            "name": "devOpsAuditingSettingsName",
```
what happens if your service receives a name different than default? perhaps is better to specify /default which is the common pattern for singletons.

eswatad replied:
In case name is not default, a not found settings error will be returned.
We didn't restrict this to default, so we can extend it to other values if needed in the future.

jorgecotillo replied:
So is your RP restricting any other name than default for now and later you'll relax this restriction? In that case, may I ask why you are forcing "default" as the name when later it can be relaxed?

jorgecotillo commented on specification/sql/resource-manager/Microsoft.Sql/preview/2020-08-01-preview/Databases.json:
```diff
@@ -481,47 +492,37 @@
-          "Pauses a database.": {
-            "$ref": "./examples/PauseDatabase.json"
+          "Failover an database": {
+            "$ref": "./examples/FailoverDatabase.json"
```
I don't see the renamed file in this PR, is it missing?

ericshape replied:
same comment as I replied in Line 472.

jorgecotillo commented on specification/sql/resource-manager/Microsoft.Sql/preview/2020-08-01-preview/Databases.json:
```diff
@@ -468,10 +464,25 @@
             "type": "string"
           },
+          {
+            "name": "replicaType",
```
Can you please make sure to add an example (or use the existing one) with this query string as well?

jorgecotillo replied:
Probably? I didn't know what algorithm git uses but I noticed that has different ones like patience, anyways, just checkout your PR and noticed that the examples are correct.

ericshape replied:
LOL~ algorithm sometimes cannot works

jorgecotillo commented on specification/sql/resource-manager/Microsoft.Sql/preview/2020-02-02-preview/Servers.json:
```diff
@@ -327,6 +382,221 @@
+            "create"
+          ]
+        },
+        "administratorLoginPassword": {
```
Is this a secret?

Make sure to add: 

"x-ms-secret": true

ericshape replied:
same comment as I replied in Line 439

jorgecotillo commented on specification/sql/resource-manager/Microsoft.Sql/preview/2020-02-02-preview/Servers.json:
```diff
@@ -327,6 +382,221 @@
+            "create"
+          ]
+        },
+        "storageKey": {
```
Is this a secret?

Make sure to add: 

"x-ms-secret": true

jorgecotillo replied:
Same for your 2020-08-01-preview api

ericshape replied:
we would try our best to catch up with the change to the latest version: 2020-08-01-preview.

jorgecotillo replied:
Thanks! Not sure if this something you can add but probably for future PRs this info can be added in the PR description.

ericshape committed:
update import export spec file.

ericshape committed:
update import export spec file.

ericshape committed:
Merge branch 'master' into fix_arm_errors

ericshape commented:
all the newly introduced breaking changes are expected. we revert DB import/export API changes in SQL DB repo: https://msdata.visualstudio.com/Database%20Systems/_git/DsMainDev/pullrequest/501089

ericshape commented:
@akning-ms it seems the doc pipeline is still broken.

akning-ms commented:
@jianyexi to help take a look Lint issue

ericshape committed:
remove import/export from v2 and v1

in the original pr: https://github.com/Azure/azure-rest-api-specs/commit/99b5b0c4914d9d485d6106fbd620ce8ef8bbea05#
the import export APIs were only added to v3.

ericshape commented:
> @jianyexi to help take a look Lint issue

lint diff is addressed. 
in previous updates, I added import/export APIs to v1 and v2 SDK incorrectly.

akning-ms commented on specification/sql/resource-manager/Microsoft.Sql/preview/2020-08-01-preview/ImportExport.json:
```diff
@@ -0,0 +1,577 @@
+          "ImportExport"
+        ],
+        "description": "Imports a bacpac into a new database.",
+        "operationId": "Databases_Import",
```
I think the previous operation id should be: ImportExport.Import_Import,  right?

akning-ms commented on specification/sql/resource-manager/Microsoft.Sql/preview/2020-02-02-preview/ImportExport.json:
```diff
@@ -0,0 +1,577 @@
+          "ImportExport"
+        ],
+        "description": "Imports a bacpac into a new database.",
+        "operationId": "Databases_Import",
```
I think the previous operation id should be: ImportExport_Import,  right?

ericshape replied:
this operation id never works in that it does not exist in SQL DB repo. 
when the original author introduced it, the operation id in SQL DB repo is "Databases_Import". I have replied your message in the internal group chat.

ericshape replied:
due to the complexity of the change. I will remove import export from this PR and you can merge it. 
Then I will create a new PR to address the import/export rollback issue.

jorgecotillo commented:
From ARM side, this PR seems ready I just left one additional comment regarding your /default endpoint.
Please try to get an approval on your breaking change.

ericshape commented:
> From ARM side, this PR seems ready I just left one additional comment regarding your /default endpoint.
> Please try to get an approval on your breaking change, since these are preview api, I think there shouldn't be any issues.

@jorgearteiro  thanks for your review! we will improve the swagger generation to address the issue you mentioned soon.

ericshape committed:
Merge branch 'fix_arm_errors' of https://github.com/ericshape/azure-rest-api-specs into fix_arm_errors

ericshape committed:
separate db import export from this PR.

ericshape committed:
Merge branch 'master' of https://github.com/ericshape/azure-rest-api-specs into fix_arm_errors

ericshape commented:
move ImportDatabase operation id from databases.json to server.json. 
The reason is that importDatabase use server API path:
/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Sql/servers/{serverName}/import
move to server.json would keep the namespace consistent.

ericshape committed:
revert importdatabases

akning-ms commented:
as ARM reviewed in PR#10994 for new adding API ServerDevopsAudit.json and this PR is adding missing api in a old version.  so removed ARM review label here, and merged this PR
